### ISSUE TITLE
dns service (192.168.64.1) not working

### ISSUE THREAD

**Issue by lethargosapatheia** [canonical/multipass#2818]
**Describe the bug**
DHCP server not working from inside the VM or from the host.

**To Reproduce**
How, and what happened?
`dig A google.com @192.168.64.1` doesn't output anything on the host and from within the virtual machine (Ubuntu 22.04) it says 'connection refused'.

**Logs**
Please provide logs from the daemon, see [accessing logs](https://multipass.run/docs/accessing-logs) on where to find them on your platform.
[multipassd.log](https://github.com/canonical/multipass/files/9971488/multipassd.log)

**Additional info**
```
 - Mac OS Monterey 12.6
```
- `multipass version`
```
multipass   1.10.1+mac
multipassd  1.10.1+mac
```
- `multipass info --all`
```
Name:           k8s
State:          Running
IPv4:           192.168.64.2
                172.17.0.1
                10.220.77.0
Release:        Ubuntu 22.04.1 LTS
Image hash:     c363ebe92173 (Ubuntu 22.04 LTS)
Load:           1.46 1.73 1.51
Disk usage:     4.3G out of 58.0G
Memory usage:   1.2G out of 7.7G
Mounts:         /Users/user/work => /root/work
                    UID map: 502:default
                    GID map: 20:default
```
- `multipass get local.driver`
```
qemu
```
**Additional context**
I've installed a virtual machine with Ubuntu 22.04 and I've installed a single-node kubernetes cluster. That's where the problem occurred initially, coredns complained about it.
For some reason, the DNS somehow worked anyway when installing various packages etc., and I'm not sure how, it can only think of the google/cloudflare dns fallback, I guess.

In case it helps:
```
multipass networks
Name     Type         Description
bridge0  bridge       Network bridge with en1, en2, en3
en0      wifi         Wi-Fi
en1      thunderbolt  Thunderbolt 1
en2      thunderbolt  Thunderbolt 2
en3      thunderbolt  Thunderbolt 3
en4      ethernet     Ethernet Adapter (en4)
en5      ethernet     Ethernet Adapter (en5)
en6      ethernet     Ethernet Adapter (en6)
en7      usb          USB 10/100/1000 LAN
```

P. S. I've started this as a bug, but I actually wanted it to report it as an issue only, because I'm not sure if it's a bug or not.

**Comment by ricab** [comment]
Hi @lethargosapatheia, Multipass does not provide DHCP or DNS servers for bridged instances. I am closing this one as it does appear to be out of scope. Please reopen if you find there is a problem on Multipass itself.

**Comment by lethargosapatheia** [comment]
Hi, @ricab. I'm a little bit confused, to be honest, because this is the exact setup I have on my other laptop where both the DHCP and the DNS services, which work on bridge interface (on the default subnet 192.168.64.0/24, which I assume is set by multipass), work perfectly.
This is how I launched the VM, so this is the default:
```
multipass launch 22.04 -d 100G -c 4 -m 8g --mount /Users/user/work:/root/work -n k8s
```
It's even more confusing for me that you say that not even the dhcp server is supposed to work, although it *does* work, as my VM gets an ip from somewhere and the DNS server is set to 192.168.64.1, which is the host's IP on that network.

Maybe I'm missing something here?

Let's be clear that this isn't bridged to the external local network, it's a virtual internal network to which only the VM and the host have access. This is the default behaviour on multipass/qemu on  mac os.

**Comment by ricab** [comment]
I am confused too. The instance you showed above has three IPs, one of which on the same network as your host, which is why I mentioned it being bridged. 

The fact that you see those IPs shows that DHCP worked, so that part looks alright. Concerning DNS, on bridged networks it is whatever is configured for that network. On the default network, DNS isn't provided by Multipass itself but by macOS. Have a look [here](https://multipass.run/docs/troubleshoot-networking#heading--dns-problems) to see if there is anything that can help figuring it out.

**Comment by lethargosapatheia** [comment]
But I haven't configured anything specially on that network, it was only what multipassed configured (or triggered). It's not *whatever*, in my opinion.

In any case, multipass stopped working completely, just like it did several months ago, when I last tried it. I thought I would give it a try again, now, but unfortunately I get these inconsistent behaviour where the virtual machines refuse to start at all and remain in an unknown state, just like I experienced before. So I'm not sure where I'm supposed to go from here.

**Comment by townsend2010** [comment]
@lethargosapatheia,

Regarding multipass not working, are you perhaps on an M1 Mac?  You may be hitting #2387.  If you are able to, turning off the macOS firewall or running it in debug mode may help here.

**Comment by lethargosapatheia** [comment]
@townsend2010 yes, I am. I also  offered feedback on an older different thread related to M1-issues. Disabling the firewall completely also doesn't help though. (I use little snitch and the built-in firewall, and I've tried disabling both)

**Comment by townsend2010** [comment]
Ok, I assume you disabled and rebooted after disabling, correct?  If so, then there is probably something else interfering with the networking then that is not clear to us.  At any rate, sorry for you troubles here 🙁

**Comment by lethargosapatheia** [comment]
@ricab returning to the initial issue, I still see the default setup as a problem, because the dhcp advertises 192.168.64.1 as the main dns server, which now indeed it seems not to work on either laptops on that ip (I guess that's expected and before it was just a weird thing?)
The weird thing is that, even though 192.168.64.1 is advertised as the main dns server, and even though I cannot query it directly (using dig, for example), the dns resolving somehow works:
```
root@k8s-2:~# resolvectl dns
Global:
Link 2 (enp0s1): 192.168.64.1 fe80::3c06:30ff:fe20:5464%65535
```
But strangely enough, tcpdump only tells me that there are dns queries being sent to 127.0.0.53:53 and that's about it. It doesn't tell me where the traffic is forwarded:
```
23:20:56.127336 lo    In  IP 127.0.0.1.48922 > 127.0.0.53.53: 20368+ [1au] AAAA? k8s.gcr.io. (39)
23:20:56.127492 lo    In  IP 127.0.0.53.53 > 127.0.0.1.48922: 63134 2/0/1 CNAME googlecode.l.googleusercontent.com., A 142.251.5.82 (103)
```
But that's related more to systemd-resolved, of course.

**Comment by lethargosapatheia** [comment]
Regarding the link you've suggested (https://multipass.run/docs/troubleshoot-networking#heading--dns-problems), this doesn't seem correct to me:
```
sudo lsof -iTCP:53 -iUDP:53 -n -P
COMMAND   PID       	USER   FD   TYPE         	DEVICE SIZE/OFF NODE NAME
mDNSRespo 191 _mdnsresponder   17u  IPv4 0xa89d451b9ea11d87  	0t0  UDP *:53
mDNSRespo 191 _mdnsresponder   25u  IPv6 0xa89d451b9ea1203f  	0t0  UDP *:53
mDNSRespo 191 _mdnsresponder   50u  IPv4 0xa89d451b9ea8b8cf  	0t0  TCP *:53 (LISTEN)
mDNSRespo 191 _mdnsresponder   55u  IPv6 0xa89d451b9e2e200f  	0t0  TCP *:53 (LISTEN)
```
In any case, this is not what I see. mDNS listens on 5353 (it's multicast dns, that's its port) and I can confirm this when running it on my laptop:
```
COMMAND   PID           USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
mDNSRespo 485 _mdnsresponder    6u  IPv4 0xeb0dabe53208468f      0t0  UDP *:5353
mDNSRespo 485 _mdnsresponder    7u  IPv6 0xeb0dabe5320849bf      0t0  UDP *:5353
```

I also think the output in your link is wrong, because I don't think mDNS uses TCP at all.

**Comment by ricab** [comment]
Hi @lethargosapatheia, I think I understand what you mean a little better now. When I saw "DHCP server" I had not understood that you were referring to the DNS server that is offered by DHCP.

`127.0.0.53` is just the interface for the *internal DNS stub resolver of systemd-resolved* in Ubuntu (have a look at the comments in `/etc/resolv.conf`). But the uplink DNS servers are still used by `systemd-resolved`. 

Do I understand correctly that you observe a `ping ubuntu.com` succeeding where the `dig` command failed? And what about `resolvectl query ubuntu.com`? Did disabling the firewall improve things for you on this front? 

Hard to say why things would be failing in your setup, but if you have a "custom stub resolver", that might be related ;) You can get some insight into what `resolved` is doing with something like:

```
$ sudo resolvectl log-level debug
$ ping -c1 ubuntu.com
$ journalctl -u systemd-resolved
```

Hopefully that will help you debug. At this point, I can't pinpoint anything Multipass itself is doing wrong, nor do I have a handle to investigate. But please let us know of your findings.

**Comment by lethargosapatheia** [comment]
> When I saw "DHCP server" I had not understood that you were referring to the DNS server that is offered by DHCP.
@ricab I'm not 100% sure if *I* understand now :), but to summarise it again: there is a DHCP server running somewhere (within the 192.168.64.0/24 network on bridge100) that offers the DNS server 192.168.64.1. So that much is clear and I'm guessing this what you expect to from a default setup on mac.

I know about systemd-resolved, I've read quite a lot about it while trying to do all sorts of things (trying to do domain-based routing among other things :) ), so yes, I know what 127.0.0.53 refers to.

So eventually I understod what was happenning in my case. I was running my own dns-resolver on my mac (which listens on 127.0.0.1:53) in order to skip the https pinning which Safari and cloudflare impose (very different topic, I know).

So because of that, when I started the multipass instance, mdnsresolver would start running on port 53, but only on IPv6. So actually through the IPv6 resolver was the virtual machine able to resolve the DNS names.

But when I started running the kubernetes cluster, coredns wasn't connecting to 127.0.0.53 (for obviously reasons, because it would have connected to itself) to forward the dns requests, but specifically to 192.168.64.1, so that meant IPv4 only. And that's why the dns requests of the kubernetes cluster didn't work. Whereas virtual machine itself (so outside the cluster) could resolve everything, because the stub-resolver would forward requests to IPv6 also. So yes, sorry about that, the whole thing was seriously affected by own setup.

The only issue that I have now is related to the idiosyncrasies of the mac firewall.

Thanks for your assistance :)
